orderer/common/cluster/commauth.go

```diff
@@ -0,0 +1,331 @@
+	return func() (*RemoteContext, error) {
+		ac.Logger.Debugf("Connecting to node: %v for channel: %v", stub.RemoteNode.NodeAddress, channel)
+
+		conn, err := ac.Connections.Connect(stub.Endpoint)
+		if err != nil {
+			ac.Logger.Warningf("Unable to obtain connection to %d(%s) (channel %s): %v", stub.ID, stub.Endpoint, channel, err)
+			return nil, err
+		}
+
+		probeConnection := func(conn *grpc.ClientConn) error {
+			connState := conn.GetState()
+			if connState == connectivity.Connecting {
+				return errors.Errorf("connection to %d(%s) is in state %s", stub.ID, stub.Endpoint, connState)
+			}
+			return nil
+		}
+
+		clusterClient := orderer.NewClusterNodeServiceClient(conn)
+		getStepClientStream := func(ctx context.Context) (StepClientStream, error) {
+			stream, err := clusterClient.Step(ctx)
+			if err != nil {
+				return nil, err
+			}
+			stepClientStream := &NodeClientStream{
+				StepClient: stream,
+			}
+			return stepClientStream, nil
+		}
+
+		workerCountReporter := workerCountReporter{
+			channel: channel,
+		}
+
+		rc := &RemoteContext{
+			Metrics:             ac.Metrics,
+			workerCountReporter: workerCountReporter,
+			Channel:             channel,
+			SendBuffSize:        ac.SendBufferSize,
+			endpoint:            stub.Endpoint,
+			Logger:              ac.Logger,
+			ProbeConn:           probeConnection,
+			conn:                conn,
+			GetStreamFunc:       getStepClientStream,
+			// Client:              clusterClient,
+			shutdownSignal:    ac.shutdownSignal,
+			SourceNodeID:      ac.NodeID,
+			DestinationNodeID: stub.ID,
+			Signer:            ac.Signer,
+		}
+		return rc, nil
+	}
+}
+
+type NodeClientStream struct {
+	StepClient orderer.ClusterNodeService_StepClient
+}
+
+func (cs *NodeClientStream) Send(request *orderer.StepRequest) error {
+	stepRequest, cerr := BuildStepRequest(request)
+	if cerr != nil {
+		return cerr
+	}
+	return cs.StepClient.Send(stepRequest)
+}
+
+func (cs *NodeClientStream) Recv() (*orderer.StepResponse, error) {
+	nodeResponse, err := cs.StepClient.Recv()
+	if err != nil {
+		return nil, err
+	}
+	return BuildStepRespone(nodeResponse)
+}
+
+func (cs *NodeClientStream) Auth(version uint32, fromId uint64, toId uint64, channel string, signer identity.Signer) error {
+	if signer == nil {
+		return errors.New("signer is nil")
+	}
+
+	timestamp, err := ptypes.TimestampProto(time.Now().UTC())
+	if err != nil {
+		return errors.Wrap(err, "failed to read timestamp")
+	}
+
+	payload := &orderer.NodeAuthRequest{
+		Version:   version,
+		Timestamp: timestamp,
+		FromId:    fromId,
+		ToId:      toId,
+		Channel:   channel,
+	}
+
+	bindingFieldsHash := GetSessionBindingHash(payload)
+
+	tlsBinding, err := GetTLSSessionBinding(cs.StepClient.Context(), bindingFieldsHash)
+	if err != nil {
+		return errors.Wrap(err, "TLSBinding failed")
+	}
+	payload.SessionBinding = tlsBinding
+
+	asnSignFields, _ := asn1.Marshal(AuthRequestSignature{
```

Comment:
We are supposed to sign over the session ID as well, see the related code from the RFC: 


> The signature is the signature that is verifiable under the initiator's public key over the ASN1 encoding of <version, timestamp, from_id, to_id, session_id, channel>.

In the code you have here, you just sign over the rest of the fields but then send the session binding (session ID) alongside, unsigned.

Reply:
This also means we don't have a unit test that checks this MITM attack that the RFC explains at length how to avoid. Can we have a UT that tests this?

Reply:
take care now. the session_binding is included now.